### PR TITLE
feat: Gatsby v4

### PR DESCRIPTION
**Pull request by tlgimenes**
## What's the purpose of this pull request?
This PR migrates the starter to use gatsby v4. Also, alongside these changes we:
- Upgrade React from 17.x to 18.x:
React was migrated without any new features being used. This should help later PRs to use the new SSR stream renderer and transitions.

- Migrates Gatsby config files from JS to TS:
Helps having `TypeScript` static type analysis on `gatsby-*` files. Also, brigs our server to TypeScript

- Add stale while revalidate headers:
Adds a huge stale-while-revalidate header to avoid re-rendering each page at every visit. This is similar to Vercel's automatic static regeneration

- Uses new JSX runtime to `automatic`:
The new JSX runtime allows us to write `.tsx` files without the `import React from "react"` statement. Also, it brings some performance improvements since spread props can be optimized. 

## How to test it?
Nothing should have changed. Just test all features as usual

## Checklist?
- [x] CHANGELOG entry added

### CODE REVIEW

File: src/pages/[...slug].tsx [hunk 1]
```diff
@@ -135,17 +156,24 @@ function Page(props: Props) {
...
         title
         description
       }
     }
+  }
+`
 
-    collection: storeCollection(id: { eq: $id }) {
+/**
+ * This query is run during SSG
```
Comment (georgebrindeiro) [inline src/pages/[...slug].tsx:172]: This should be SSR, right?

File: package.json [hunk 2]
```diff
@@ -30,35 +30,36 @@
     "@faststore/api": "^1.6.13",
     "@faststore/sdk": "^1.6.13",
     "@faststore/ui": "^1.6.13",
-    "@vtex/gatsby-plugin-nginx": "^1.6.12",
     "@vtex/gatsby-plugin-thumbor": "^1.6.12",
-    "@vtex/gatsby-source-store": "^1.6.12",
```
Comment (georgebrindeiro) [inline package.json:35]: Why is @vtex/gatsby-source-store no longer needed?

File: gatsby-config.ts [hunk 3]
```diff
@@ -1,22 +1,24 @@
...
-const config = require('./store.config')
+import config from './store.config'
+
+dotenv.config({ path: 'vtex.env' })
 
 const {
   NODE_ENV,
   URL = config.storeUrl,
   DEPLOY_PRIME_URL = URL,
   CONTEXT: ENV = NODE_ENV,
```
Comment (filipewl) [inline gatsby-config.ts:14]: While we're here, do you think this can be removed as well? It's not being used within this file.

CleanShot 2022-04-04 at 15 43 54

```suggestion
```

File: gatsby-node.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import type { GatsbyNode } from 'gatsby'
+
+import { apiSchema } from './src/server'
+
+export const onPreInit: GatsbyNode['onPreInit'] = async ({ reporter }) => {
+  reporter.info('Copying Partytown Files')
+
+  await copyLibFiles(path.resolve('./public/~partytown'))
+}
+
```
Comment (filipewl) [inline gatsby-node.ts:13]: I think this bit got removed in #433.

cc @igorbrasileiro 

```suggestion
import type { GatsbyNode } from 'gatsby'

import { apiSchema } from './src/server'

```

File: gatsby-config.ts (hunk 3 above)
Comment (tlgimenes) [reply gatsby-config.ts:14]: we use for `ENV`, that is used. for `isProduction` etc..

File: gatsby-node.ts (hunk 4 above)
Comment (tlgimenes) [reply gatsby-node.ts:13]: nice catch!

File: vtex.env [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # configures Gatsby
 GATSBY_CPU_COUNT=4
 ENABLE_GATSBY_REFRESH_ENDPOINT=true
+GATSBY_QUERY_ON_DEMAND_LOADING_INDICATOR=false
```
Comment (filipewl) [inline vtex.env:4]: Is this being used anywhere?

CleanShot 2022-04-04 at 16 10 16

```suggestion
```

Comment (tlgimenes) [reply vtex.env:4]: Yes, this is used by Gatsby internally. https://github.com/gatsbyjs/gatsby/blob/ee47e27c843910bf08887f270b831bbaeb6ab538/packages/gatsby/cache-dir/app.js#L163

File: gatsby-config.ts (hunk 3 above)
Comment (filipewl) [reply gatsby-config.ts:14]: I know about `ENV`, but I don't see `CONTEXT` being **directly** used in the file so I thought it wasn't being used.

File: cypress.json [hunk 6]
```diff
@@ -5,6 +5,6 @@
   "video": true,
   "retries": 2,
   "chromeWebSecurity": false,
-  "viewportWidth": 1000,
-  "viewportHeight": 600
+  "viewportWidth": 360,
+  "viewportHeight": 640
```
Comment (filipewl) [inline cypress.json]: I liked this change at first (LH's Moto G4 dimensions), but we should remember that the integration tests from WebOps CI won't use these values and will keep using 1000 x 600 so there could be some minor results in local/CI runs. 

Ideally, there shouldn't be many differences, but apparently, there are some. I think the manual position adjustments we needed to make to the infinite scroll tests were considering the 1000 x 600 in mind.

**Review by filipewl — APPROVED** [review]
Awesome work! LGTM!

File: src/components/common/Navbar/Navbar.tsx [hunk 7]
```diff
@@ -13,37 +13,38 @@ import { mark } from 'src/sdk/tests/mark'
...
+  },
+  {
+    name: 'Computer and Software',
+    href: '/computer-and-software',
+  },
+  {
+    name: 'Technology',
+    href: '/technology',
+  },
+]
```
Comment (eduardoformiga) [inline src/components/common/Navbar/Navbar.tsx:38]: out of curiosity: why these mocked collections? should it come as props?

File: src/components/sections/Section/Section.tsx [hunk 8]
```diff
@@ -1,12 +1,9 @@
-import React from 'react'
-import type { PropsWithChildren } from 'react'
+import type { DetailedHTMLProps } from 'react'
```
Comment (eduardoformiga) [inline src/components/sections/Section/Section.tsx:1]:
```suggestion
import type { DetailedHTMLProps, HTMLAttributes } from 'react'
```

File: src/components/sections/Section/Section.tsx [hunk 9]
```diff
@@ -1,12 +1,9 @@
...
 
-interface Props {
-  className?: string
-}
+type Props = DetailedHTMLProps<React.HTMLAttributes<HTMLElement>, HTMLElement>
```
Comment (eduardoformiga) [inline src/components/sections/Section/Section.tsx:3]:
```suggestion
type Props = DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
```

File: src/pages/[...slug].tsx [hunk 10]
```diff
@@ -162,6 +189,37 @@ export const query = graphql`
...
+    const { data } = await execute({
+      operationName: querySSR,
+      variables: { slug },
+    })
+
+    return {
+      status: 200,
+      props: data ?? {},
+      headers: {
+        'cache-control': 'public, max-age=0, stale-while-revalidate=31536000',
```
Comment (eduardoformiga) [inline src/pages/[...slug].tsx:208]: what does the `31536000` number mean? What if we extract to a named constant or add a comment?

File: src/pages/[slug]/p.tsx [hunk 11]
```diff
@@ -174,6 +198,40 @@ export const querySSG = graphql`
...
+    const { data } = await execute({
+      operationName: querySSR,
+      variables: { id },
+    })
+
+    return {
+      status: 200,
+      props: data ?? {},
+      headers: {
+        'cache-control': 'public, max-age=0, stale-while-revalidate=31536000',
```
Comment (eduardoformiga) [inline src/pages/[slug]/p.tsx:219]: the same here for the `31536000` number

File: src/pages/[slug]/p.tsx [hunk 12]
```diff
@@ -1,45 +1,59 @@
...
+    if (notFound) {
+      window.location.href = `/404/?from=${encodeURIComponent(
+        window.location.pathname
+      )}`
+    }
+  }, [notFound])
+
+  if (notFound) {
+    return null
+  }
```
Comment (eduardoformiga) [inline src/pages/[slug]/p.tsx:55]: just thinking outloud: is it possible to redirect from the `getServerData` instead?

**Review by eduardoformiga — APPROVED** [review]
Awesome job! 🚀 

From my point of view, everything is working as expected! I left some minor comments.

**Review by georgebrindeiro-vtex — APPROVED** [review]
LGTM

File: src/pages/[...slug].tsx (hunk 10 above)
Comment (tlgimenes) [reply src/pages/[...slug].tsx:208]: one year in seconds

File: src/pages/[slug]/p.tsx (hunk 12 above)
Comment (tlgimenes) [reply src/pages/[slug]/p.tsx:55]: Yes, we can do it later on

File: src/components/common/Navbar/Navbar.tsx (hunk 7 above)
Comment (tlgimenes) [reply src/components/common/Navbar/Navbar.tsx:38]: we don't have those allCategories queries anymore :(

File: src/components/ThirdPartyScripts/vtex.tsx [hunk 13]
```diff
@@ -1,10 +1,7 @@
-import React from 'react'
-
 function VTEX() {
   return (
     <>
       <script
-        key="vtexrc.js-init"
```
Comment (igorbrasileiro) [inline src/components/ThirdPartyScripts/vtex.tsx:7]: Why did you remove them? I know those scripts won't be added or removed, but it helps React understand who those tags are.

File: src/components/sections/Section/Section.tsx [hunk 14]
```diff
@@ -1,12 +1,9 @@
-import React from 'react'
-import type { PropsWithChildren } from 'react'
+import type { DetailedHTMLProps, HTMLAttributes } from 'react'
 
-interface Props {
-  className?: string
-}
+type Props = DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>
```
Comment (igorbrasileiro) [inline src/components/sections/Section/Section.tsx:3]:
```suggestion
interface Props extends DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement> {}
```

File: src/components/ui/Button/ButtonSignIn.tsx [hunk 15]
```diff
@@ -1,9 +1,9 @@
-import React from 'react'
 import Icon from 'src/components/ui/Icon'
 import usePersonQuery from 'src/sdk/person/usePersonQuery'
 import { ButtonLink } from 'src/components/ui/Button'
+import type { FC } from 'react'
 
-const ButtonSignIn: React.FC = () => {
+const ButtonSignIn: FC = () => {
```
Comment (igorbrasileiro) [inline src/components/ui/Button/ButtonSignIn.tsx:6]: We already know that we should avoid using FC components. Can you remove this component type?

**Review by igorbrasileiro — COMMENTED** [review]
Great job, left some comments.

**Comment by igorbrasileiro** [comment]
This PR decreases a lot the store performance.

**Comment by igorbrasileiro** [comment]
This PR decreases a lot the page performance 
Page result - https://base.vtex.app/
| Metric | Mean | Standard deviation |
|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 |
| First Contentful Paint (FCP) | 1358.32 | 0.61 |
| First Contentful Paint 3g (FCP) | 2587.96 | 76483.96 |
| Largest Contentful Paint (LCP) | 1891.42 | 100171.66 |
| Time to Interactive (TTI) | 2877.06 | 9002.24 |
| Total Blocking Time (TBT) | 378.90 | 1975.89 |
| Performance score | 0.896 | 0.000084 |
| JavaScript Execution Time | 1309.42 | 72.96 |
| Speed Index | 1417.94 | 1462.94 |
Page result - https://sfj-443ad2a--base.preview.vtex.app/
| Metric | Mean | Standard deviation |
|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 |
| First Contentful Paint (FCP) | 1500.84 | 46751.09 |
| First Contentful Paint 3g (FCP) | 3098.65 | 82563.25 |
| Largest Contentful Paint (LCP) | 2158.28 | 121381.78 |
| Time to Interactive (TTI) | 3155.25 | 144178.69 |
| Total Blocking Time (TBT) | 621.76 | 84977.64 |
| Performance score | 0.836 | 0.008213 |
| JavaScript Execution Time | 1550.60 | 183877.18 |
| Speed Index | 1578.10 | 12305.28 |

**Comment by igorbrasileiro** [comment]
Page result - https://base.vtex.app/licensed-cotton-hat-licensed-2791588/p
| Metric | Mean | Standard deviation |
|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 |
| First Contentful Paint (FCP) | 1500.80 | 0.05 |
| First Contentful Paint 3g (FCP) | 3135.80 | 22208.37 |
| Largest Contentful Paint (LCP) | 2308.77 | 10396.01 |
| Time to Interactive (TTI) | 2850.52 | 52359.39 |
| Total Blocking Time (TBT) | 344.75 | 17380.69 |
| Performance score | 0.895 | 0.002475 |
| JavaScript Execution Time | 1181.93 | 60761.36 |
| Speed Index | 1648.94 | 12634.27 |

Page result - https://sfj-443ad2a--base.preview.vtex.app/licensed-cotton-hat-licensed-2791588/p
| Metric | Mean | Standard deviation |
|--------|--------|--------|
| Cumulative Layout shift (CLS) | 0.00 | 0.00 |
| First Contentful Paint (FCP) | 1381.32 | 48507.16 |
| First Contentful Paint 3g (FCP) | 2941.32 | 86276.20 |
| Largest Contentful Paint (LCP) | 1711.32 | 5369.08 |
| Time to Interactive (TTI) | 2964.92 | 2828.99 |
| Total Blocking Time (TBT) | 469.80 | 3630.96 |
| Performance score | 0.872 | 0.000216 |
| JavaScript Execution Time | 1333.69 | 78.66 |
| Speed Index | 1656.17 | 160032.24 |

File: .eslintrc [hunk 16]
```diff
@@ -2,7 +2,13 @@
...
   "root": true,
   "rules": {
-    "react/prop-types": "off"
+    "react/prop-types": "off",
+    "react/react-in-jsx-scope": "off"
+  },
+  "settings": {
+    "react": {
+      "version": "detect"
+    }
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply .eslintrc:11]: no, this was introduced on #373

File: gatsby-config.ts [hunk 17]
```diff
@@ -112,55 +110,6 @@ module.exports = {
         reportFilename: 'bundle-analyser.html',
       },
     },
-    {
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply gatsby-config.ts:115]: yeah, we can remove it now

File: gatsby-ssr.tsx [hunk 18]
```diff
@@ -47,15 +48,15 @@ export const onRenderBody = ({ setHeadComponents }) => {
...
+export const onPreRenderHTML: GatsbySSR['onPreRenderHTML'] = ({
   getHeadComponents,
   replaceHeadComponents,
 }) => {
   if (process.env.NODE_ENV !== 'production') {
     return
   }
 
-  const transformedHeadComponents = getHeadComponents().map((node) => {
+  const transformedHeadComponents = getHeadComponents().map((node: any) => {
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply gatsby-ssr.tsx:59]: it breaks ts

File: src/components/ThirdPartyScripts/ThirdPartyScripts.tsx [hunk 19]
```diff
@@ -12,19 +11,18 @@ const gtmContainerId = storeConfig.analytics?.gtmContainerId
 const includeGTM = typeof gtmContainerId === 'string'
 const includeVTEX = storeConfig.platform === 'vtex'
 
-function ThirdPartyScripts() {
-  if (process.env.NODE_ENV === 'development' && !includeGTM) {
-    console.warn(
-      'No GTM container id found. Check the analytics section on your store.config.js file for enhanced observability of your store.'
-    )
-  }
+if (process.env.NODE_ENV === 'development' && !includeGTM) {
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply src/components/ThirdPartyScripts/ThirdPartyScripts.tsx:14]: less computing while rendering

File: src/components/ThirdPartyScripts/vtex.tsx (hunk 13 above)
Comment (tlgimenes) [reply src/components/ThirdPartyScripts/vtex.tsx:7]: why?

File: src/components/sections/Section/Section.tsx (hunk 14 above)
Comment (tlgimenes) [reply src/components/sections/Section/Section.tsx:3]: eslint prefers types

File: gatsby-ssr.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (victorhmp) [reply gatsby-ssr.tsx:59]: https://github.com/vtex-sites/base.store/pull/332/files#diff-6947b63960353c68fcbfbfca1a2c4f803766c6b36c3a67408261c24fabc9abcbR14

File: gatsby-config.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (emersonlaurentino) [reply gatsby-config.ts:14]: you need to change the usage on plugin `gatsby-plugin-robots-txt`

https://github.com/vtex-sites/base.store/pull/207/files#diff-26b5c624a22b9f24d3e683f7ee5b69788ee6b166a655b7423b6a4c5a1ef6f0a1R52

File: src/server/index.ts [hunk 20]
```diff
@@ -1,38 +1,53 @@
 /* eslint-disable react-hooks/rules-of-hooks */
-const {
+/**
+ * Polyfill for dataloader.
+ * TODO: Remove it once this is fixed: https://github.com/graphql/dataloader/issues/249
```
Comment (danzanzini) [inline src/server/index.ts:4]: The issue above is already closed, it seems to be fixed here: https://github.com/graphql/dataloader/pull/293

File: src/pages/index.tsx [hunk 21]
```diff
@@ -104,7 +103,7 @@ function Page(props: Props) {
   )
 }
 
-export const query = graphql`
+export const querySSG = graphql`
```
Comment (danzanzini) [inline src/pages/index.tsx:106]: How can we know if the query runs at SSG or not?

File: src/pages/[slug]/p.tsx [hunk 22]
```diff
@@ -174,6 +184,53 @@ export const querySSG = graphql`
   }
 `
 
+export const getServerData = async ({
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply src/pages/[slug]/p.tsx:181]: not really. People may want to add different checks. Let's leave it open and see how people use it

File: src/pages/index.tsx (hunk 21 above)
Comment (tlgimenes) [reply src/pages/index.tsx:106]: if it has `graphql` it runs on SSG. If it has a `gql` it runs on SSR or client side

File: src/server/index.ts (hunk 20 above)
Comment (tlgimenes) [reply src/server/index.ts:4]: nice, we can check it on a later pr

File: gatsby-config.ts (hunk 3 above)
Comment (filipewl) [reply gatsby-config.ts]: Oh, my bad! I didn't realize `CONTEXT` here was actually being renamed to `ENV`! 🤦‍♂️

File: src/server/index.ts [hunk 23]
```diff
@@ -1,38 +1,52 @@
...
+import { useParserCache } from '@envelop/parser-cache'
+import { useValidationCache } from '@envelop/validation-cache'
+import { getContextFactory, getSchema } from '@faststore/api'
+import { GraphQLError } from 'graphql'
+import type { FormatErrorHandler } from '@envelop/core'
+import type { Options as APIOptions } from '@faststore/api'
+import type { GatsbyFunctionRequest } from 'gatsby'
+
+import persisted from '../../@generated/graphql/persisted.json'
+import storeConfig from '../../store.config'
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply src/server/index.ts:24]: not really. lighthouse.js uses this file and it's a js file. We either move ligthouse config elsewhere or keep store.config js

File: tsconfig.json [hunk 24]
```diff
@@ -14,6 +16,12 @@
     },
     "forceConsistentCasingInFileNames": true
   },
-  "include": ["src/**/*.ts", "src/**/*.tsx", "@generated/**/*.ts"],
+  "include": [
+    "src/**/*.ts",
+    "src/**/*.tsx",
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply tsconfig.json:21]: somehow, when I do this, tsc stops working
